Add manual page on variants (#1930)
informalsystems/apalache: docs/src/lang/variants.md

# Variants

**Warning:** This feature is currently under testing. To enable variants
support in Apalache, pass the option `--features=rows`.

> [[Back to all operators]](./apalache-extensions.md)

Author: We have the back links in other sections too. I find them useful, as it is easy to get lost in the manual structure.

[Variants][] (also called *tagged unions*) are useful, when you want to combine
values of different shapes in a single set or a sequence.

**Idiomatic tagged unions in untyped TLA+.** In untyped TLA+, one can simply
pack a value into a record and add an additional tag field. For instance, we
could create a set that contains two records of different shapes:

```tla
ApplesAndOranges == {
    [ tag |-> "Apple", color |-> "red" ],
    [ tag |-> "Orange", seedless |-> TRUE ]
  }
```

Now we can unpack the elements of `ApplesAndOranges` based on their tag:

```tla
  \E e \in ApplesAndOranges:
    /\ e.tag = "Apple"
    /\ e.color /= "green"
```

This idiom is quite common in untyped TLA+. [Tagged unions in Paxos][] is
probably the most illuminating example of this idiom. Unfortunately, it is way
too easy to make a typo in the tag name, since it is a string, or simply access

> a wrong record without checking its tag.

Reviewer: Unfortunately, backticks don't work well in "suggest" mode...

**Variants module.** Apalache formalizes the above idiom in the module
[Variants.tla][]. Due to the type checker, the users get a type error, when
they access a wrong value. Additionally, the default implementation raises an
error in TLC, when a variant is used incorrectly.

**Immutability**. As variants are immutable.

**Construction.** An instance of a variant can be constructed via the operator
`Variant`:

```tla
Variant("Apple", "red")
```

If we just construct a variant like in the example above, it will be assigned
a parametric variant type:

```
Apple(Str) | a
```

In this type, we know that whenever a value is tagged with "Apple" it should be
of the string type. However, we know nothing about other options. Most of the
time, we want to define variants that are sealed, that is, we know all
available options. In the Apalache terms, it means that our type would look

> like follows:

Reviewer 2: ```suggestion
like the following:
```

> available options. In the Apalache terms, it means that our type would look
> like follows:

Reviewer 2: ```suggestion
available options. Suppose we wanted to reason about different kinds of fruit, 
but wanted to limit our model to only comparing apples and oranges.
In Apalache, the type for a value that could be either an apple or an orange, but nothing else, 
would be as follows:
```

```
Apple(Str) | Orange(Bool)
```

To do this, we can introduce variants together with user-defined
constructors for each option right away:

```tla

> \* @typeAlias: FRUIT = Apple(Str) | Orange(Bool);
> \*
> \* @type: Str => FRUIT;
> Apple(color) == Variant("Apple", color)
> 
> \* @type: Bool => FRUIT;
> Orange(seedless) == Variant("Orange", seedless)

Reviewer 3: Nitpicking here, but do you consider this interleaved `\*` idiomatic?
I've seen it used below as well, but I find the grouping confusing, because then `Orange` is offset by an empty line.

```suggestion
\* @typeAlias: FRUIT = Apple(Str) | Orange(Bool);

\* @type: Str => FRUIT;
Apple(color) == Variant("Apple", color)

\* @type: Bool => FRUIT;
Orange(seedless) == Variant("Orange", seedless)
```

Author: Makes sense, though we have to make this works with SANY.

```

Now we can naturally construct apples and orange as follows:

```tla
Apple("red")
Orange(TRUE)
```

> Note that we have lost the field names in the process. If we want to keep
> the names, we should use records instead of basic values:
> 
> ```tla
> \* @typeAlias: DRINK =
> \*     Water({ sparkling: Bool })
> \*   | Beer({ malt: Str, strength: Int });
> \*
> \* @type: Bool => DRINK;
> Water(sparkling) == Variant("Water", [ sparkling |-> sparkling ])
> 
> \* @type: (Str, Int) => DRINK;
> Beer(malt, strength) == Variant("Beer", [ malt |-> malt, strength |-> strength ])
> ```

Reviewer 2: Framing this as  "losing the field names" seems a bit confusing to me. Could it be framed just as: "the value in a variant can be a record"? E.g.,

```suggestion
Variants can wrap records, for when we want to represent compound data with named fields:

```tla
\* @typeAlias: DRINK =
\*     Water({ sparkling: Bool })
\*   | Beer({ malt: Str, strength: Int });
\*
\* @type: Bool => DRINK;
Water(sparkling) == Variant("Water", [ sparkling |-> sparkling ])

\* @type: (Str, Int) => DRINK;
Beer(malt, strength) == Variant("Beer", [ malt |-> malt, strength |-> strength ])
```
```

operators, presented below.

**Filtering by tag name.** Following the idiomatic use of tagged unions in
untyped TLA+, we can filter a set of variants:

```tla
LET Drinks == { Water(TRUE), Water(FALSE), Beer("Radler", 2) } IN
\E d \in VariantFilter("Beer", Drink):
    d.strength < 3
```

We believe that `VariantFilter` is the most commonly used way to partition a
set of variants. Note that `VariantFilter` transforms a set of variants into a
set of values (that correspond to the associated tag name).

**Type-safe get.** Sometimes, we do have just a value that does not belong to a
set, so we cannot use `VariantFilter` directly. In this case, we can use
`VariantGetOrElse`:

```tla
LET water == Water(TRUE) IN
VariantGetOrElse("Beer", water,
                 [ malt |-> "Non-alcoholic", strength |-> 0])).strength
```

In the above example, we unpack `water` by using the tag name `"Beer"`.  Since
`water` is actually tagged with `"Water"`, the operator falls back to the
default case and returns the record `[ malt |-> "Non-alcoholic", strength |->
0]`.

**Type-unsafe get.** Sometimes, using `VariantFilter` and `VariantGetOrElse`

> is a nuisance, when we know exactly the value type. In this case, we can bypass

Reviewer: ```suggestion
is a nuisance, when we know the exact value type. In this case, we can bypass
```

the type checker and get the value notwithstanding the tag:

```tla
LET drink == ... IN
LET nonFree ==
    IF VariantTag(drink) = "Water"
    THEN VariantGetUnsafe("Water", drink).sparkling
    ELSE VariantGetUnsafe("Beer", drink).strength > 0
IN
...
```

> In general, you should avoid using `VariantGetUnsafe`, as it is type unsafe :-)

Author: Right. It will produce no error, just a value whose value we do not know.

> In general, you should avoid using `VariantGetUnsafe`, as it is type unsafe :-)

Author: Added an example.

----------------------------------------------------------------------------

## Operators

----------------------------------------------------------------------------

<a name="variantCtor"></a>
### Variant constructor

**Notation:** `Variant(tagName, associatedValue)`

**LaTeX notation:** same

**Arguments:** Two arguments: the tag name (a string literal) and a value
(a TLA+ expression).

**Apalache type:** `(Str, a) => tagName(a) | b `, for some types `a` and `b`.

> Note that `tagName` is an identifier in this notation.

Reviewer: Take some time to explain where `b` comes from in all these examples. A new reader will likely be confused by it.

**Effect:** The variant constructor returns a new value that keeps the
pair `(tagName, associatedValue)`.

**Determinism:** Deterministic.

**Errors:** No errors.

**Example in TLA+:**

```tla
\* @typeAlias: DRINK =
\*     Water({ sparkling: Bool })
\*   | Beer({ malt: Str, strength: Int });
\*
\* @type: Bool => DRINK;
Water(sparkling) == Variant("Water", [ sparkling |-> sparkling ])

\* @type: (Str, Int) => DRINK;
Beer(malt, strength) == Variant("Beer", [ malt |-> malt, strength |-> strength ])
```

----------------------------------------------------------------------------

<a name="variantTag"></a>
### Variant tag

> **Notation:** `VariantTag(variant)`

Reviewer 3: Looks like we don't support this in `Variants.tla`? Should we mention that?

**LaTeX notation:** same

**Arguments:** One argument: a variant constructed via `Variant`.

> **Apalache type:** `(tagName(a) | b => Str)`, for some types `a`
> and `b`. Note that `tagName` is an identifier in this notation.

Reviewer 3: Is the grouping here correct? I'd expect

```suggestion
**Apalache type:** `(tagName(a) | b) => Str`, for some types `a`
and `b`. Note that `tagName` is an identifier in this notation.
```

**Determinism:** Deterministic.

**Errors:** No errors.

**Example in TLA+:**

```tla
VariantTag(Variant("Water", [ sparkling |-> sparkling ])) = "Water"
```

----------------------------------------------------------------------------

<a name="variantFilter"></a>
### Variant filter

**Notation:** `VariantFilter(tagName, set)`

**LaTeX notation:** same

**Arguments:** Two arguments: the tag name (a string literal) and a set of
variants (a TLA+ expression).

**Apalache type:** `(Str, Set(tagName(a) | b)) => Set(a)`, for some types `a`
and `b`. Note that `tagName` is an identifier in this notation.

**Effect:** The variant filter keeps the set elements that are tagged with
`tagName`. It removes the tags from these elements and produces the set of
values that were packed with `Variant`.

**Determinism:** Deterministic.

**Errors:** No errors.

**Example in TLA+:**

```tla
\* @typeAlias: DRINK =
\*     Water({ sparkling: Bool })
\*   | Beer({ malt: Str, strength: Int });
\*
\* @type: Bool => DRINK;
Water(sparkling) == Variant("Water", [ sparkling |-> sparkling ])

\* @type: (Str, Int) => DRINK;
Beer(malt, strength) == Variant("Beer", [ malt |-> malt, strength |-> strength ])

> LET Drinks == { Water(TRUE), Water(FALSE), Beer("Radler", 2) } IN
> \E d \in VariantFilter("Beer", Drink):

Reviewer 3: ```suggestion
LET Drinks == { Water(TRUE), Water(FALSE), Beer("Radler", 2) } IN
\E d \in VariantFilter("Beer", Drinks):
```

**Errors:** No errors.

**Example in TLA+:**

```tla
\* @typeAlias: DRINK =
\*     Water({ sparkling: Bool })
\*   | Beer({ malt: Str, strength: Int });
\*
\* @type: Bool => DRINK;
Water(sparkling) == Variant("Water", [ sparkling |-> sparkling ])

\* @type: (Str, Int) => DRINK;
Beer(malt, strength) == Variant("Beer", [ malt |-> malt, strength |-> strength ])

LET water == Water(TRUE) IN
VariantGetOrElse("Beer", water,
                 [ malt |-> "Non-alcoholic", strength |-> 0])).strength
```

----------------------------------------------------------------------------

<a name="variantGetUnsafe"></a>
### Unpacking a variant unsafely

**Notation:** `VariantGetUnsafe(tagName, variant)`

**LaTeX notation:** same

**Arguments:** Two arguments: the tag name (a string literal) and a variant
constructed via `Variant`.

**Apalache type:** `(Str, tagName(a) | b) => a`, for some types `a` and
`b`. Note that `tagName` is an identifier in this notation.

**Effect:** The operator `VariantGetUnsafe` unconditionally returns some value
that is compatible with the type of values tagged with `tagName`. If `variant`
is tagged with `tagName`, the returned value is the value that was wrapped via

> the `Variant` constructor. Otherwise, it is some value of proper type. As such,

Reviewer: ```suggestion
the `Variant` constructor. Otherwise, it is some arbitrary value of proper type. As such,
```
